src/Microsoft.Diagnostics.Monitoring.Egress.AzureStorage/AzureBlobEgressEndpoint.cs

```diff
@@ -51,21 +37,29 @@ public AzureBlobEgressEndpoint(AzureBlobEgressEndpointOptions settings)
             BlobContainerClient containerClient = serviceClient.GetBlobContainerClient(_settings.ContainerName);
             await containerClient.CreateIfNotExistsAsync(cancellationToken: token);
 
-            string blobName = fileName;
+            string blobName = name;
             if (!string.IsNullOrEmpty(_settings.BlobDirectory))
             {
-                blobName = string.Concat(_settings.BlobDirectory, "/", fileName);
+                blobName = string.Concat(_settings.BlobDirectory, "/", name);
+            }
+
+            BlockBlobClient blobClient = containerClient.GetBlockBlobClient(blobName);
+
+            // Write blob content
+            using (Stream blobStream = await blobClient.OpenWriteAsync(overwrite: true, cancellationToken: token))
```

Comment:
Some considerations for this:
- Do we want to upload as fast as possible? For example, can this Api take advantage of High-throughput transfer? 
- If we allowed the egress Api to take the temporary file instead of a Stream, would this allow us to do more interesting things like transferring blocks in parallel, or fail transfer and pick up where we left off? I'm not saying we need those things right now, but we may be limiting our abilities in the future.

Reply:
What's "High-throughput transfer"?
I would not recommend going via disk. Attached disks are slow and will most likely become a bottleneck.
You are correct, however, that this approach cannot take advantage of parallel uploads and cannot do retries/restarts.

Reply:
For dump collection, there is an intermediate temp file (I think there's a separate thread/issue regarding creation of streamed dumps but that's not something we would have for .Net 5).

https://azure.microsoft.com/en-us/blog/high-throughput-with-azure-blob-storage/

Reply:
The BlockBlobClient.OpenWriteAsync method returns BlockBlobWriteStream which does incrementally create blocks and uploads them with the same HTTP API but does them synchronously.

Whether the egress endpoint uses what's provided by the library as-is or does parallel upload feels like an implementation detail that shouldn't be impacted by the current egress API surface. The stream the endpoint hands out to the callback action needs to be completely written to when the callback action completes, so the endpoint should be able to do whatever it wants with it regarding parallel transfers to retries.

Reply:
If I switch back to BlobClient.UploadAsync, then concurrency can be specified: https://docs.microsoft.com/en-us/dotnet/api/azure.storage.storagetransferoptions.maximumconcurrency?view=azure-dotnet

So it would be a trade off having to have the stream already point to data rather than getting a writable stream from the library. Alternatively, we could write our own stream implementation that does concurrent block puts and one block list put during the final flush. Could also allow rewrite of blocks if we really want to get fancy (the dump creation algorithm does rewrites on the stream and would be awesome if we could support that without having it write to MemoryStream of FileStream as an intermediate step).